### PR TITLE
Only show timeslots >= now + reservation lead time

### PR DESCRIPTION
**Pull request by ryanmitchell**
Should fix: https://github.com/tastyigniter/TastyIgniter/issues/617

### CODE REVIEW

File: components/Booking.php [hunk 1]
```diff
@@ -235,15 +235,19 @@ public function getTimeSlots()
...
-                'time' => $dateTime->isoFormat(lang('system::lang.moment.time_format')),
-                'fullyBooked' => $this->manager->isFullyBookedOn($dateTime, input('guest', $this->property('minGuestSize'))),
-            ];
+            if ($dateTime >= $nowPlusLeadTime)
+            {
+                $result[] = (object)[
+                    'index' => $index++,
+                    'isSelected' => $dateTime->format('H:i') == $selectedTime->format('H:i'),
+                    'rawTime' => $dateTime->format('H:i'),
+                    'time' => $dateTime->isoFormat('HH:mm'),
```
Comment (sampoyigi) [inline components/Booking.php:247]: Did you miss the lang format `$dateTime->isoFormat(lang('system::lang.moment.time_format'))`?

File: components/Booking.php [hunk 2]
```diff
@@ -235,15 +235,19 @@ public function getTimeSlots()
...
             $dateTime = $selectedDate->copy()->setTimeFromTimeString($date->format('H:i'));
-            $result[] = (object)[
-                'index' => $index++,
-                'isSelected' => $dateTime->format('H:i') == $selectedTime->format('H:i'),
-                'rawTime' => $dateTime->format('H:i'),
-                'time' => $dateTime->isoFormat(lang('system::lang.moment.time_format')),
-                'fullyBooked' => $this->manager->isFullyBookedOn($dateTime, input('guest', $this->property('minGuestSize'))),
-            ];
+            if ($dateTime >= $nowPlusLeadTime)
+            {
```
Comment (sampoyigi) [inline components/Booking.php:242]: Should be on the same line

**Review by sampoyigi — CHANGES_REQUESTED** [review]

File: components/Booking.php (hunk 1 above)
Comment (ryanmitchell) [reply components/Booking.php]: Yep good spot - I've updated it

File: components/Booking.php (hunk 2 above)
Comment (ryanmitchell) [reply components/Booking.php]: Changed!

**Comment by ryanmitchell** [comment]
Yep guest onchange is not needed with the approach we are now taking

**Review by sampoyigi — APPROVED** [review]

**Comment by sampoyigi** [comment]
Please check I haven't broken anything ;)

**Comment by muhammadsr** [comment]
Wy are we showing timeslots >= now + reservation lead time?

Shouldn't be timeslots >= now + reservation time interval?

If it's 1:00 pm right now and the restaurant has 15 min for a reservation time interval and 60 min for reservation lead time, you don't want the next available timeslot to be 2:00pm. You want it to be 01:15 pm

**Comment by sampoyigi** [comment]
The lead time is only used to determine how long it takes to prepare an ASAP order, while the interval is used mainly to generate the timeslots for later times. 

For example, the current time for your customer is 10 am and they want their order delivered/collected ASAP, we add the lead time to the current time to estimate when the order will be delivered or can be collected. On the other hand say they want the order for a later time, the lead time doesn't apply as they can explicitly choose when they want their order using the timeslots.

**Comment by muhammadsr** [comment]
@sampoyigi You are talking about orders but I was talking about reservations. The "Reservation Lead Time" is also for orders?

**Comment by sampoyigi** [comment]
Ah, my bad. IIRC should be the same logic for orders and reservations

**Comment by ryanmitchell** [comment]
> Shouldn't be timeslots >= now + reservation time interval?

if this is what you want just set the lead time to zero?

**Comment by muhammadsr** [comment]
> > Shouldn't be timeslots >= now + reservation time interval?
> 
> if this is what you want just set the lead time to zero?

@ryanmitchell This does not make sense to me. If we set the Reservation Lead Time to 0, that means customers can book multiple reservations for the same table with timeslots that are apart by the value of "Reservation Time Interval". 

Screen Shot 2021-02-10 at 2 52 54 PM
 
For example, If have the following parameters:
1. Restaurant where customers dine in for one hour 
2. it's 02:55 pm right now
3. The "Reservation Time Interval" is 15 mins
4.  We have one table

we will have the following timeslots: 
- 03:00 pm
- 03:15 pm
- 03:30 pm

A customer can book at 03:00 pm and another customer can book at 03:15 but that is not what we want because customers dine in for 1 hour on average. So, to accomplish this using the "0" "Reservation Lead Time", we will need to adjust the "Reservation Time Interval" to be 60 mins. Therefore, our timeslots will look like this now:

-03:00 pm
-04:00 pm
-05:00 pm

However, we also do not want that What if a customer wants to reserve at 03:30 pm? 

You see the issue here?

**Comment by ryanmitchell** [comment]
Yep I see what you are saying.

@sampoyigi we would amend this with the fields that exist, but maybe its worth adding a new field for reservations... so we have interval, lead time, stay time - with stay time being the amount of time they stay at a table? At the moment we're using lead time and stay time for the same function, which shouldnt really be the case.

**Comment by sampoyigi** [comment]
I think we already have a stay time field on reservations?

**Comment by ryanmitchell** [comment]
Screenshot 2021-02-11 at 07 50 51
These are the fields on my install... we're using lead time as stay time at the moment.

**Comment by sampoyigi** [comment]
Alright, makes sense to add a new stay time field then.

**Comment by ryanmitchell** [comment]
Opened a PR here https://github.com/tastyigniter/ti-ext-reservation/pull/12 - lets move the conversation over there :)

**Pull request merged**